libipl/p10/ipl0.C

```diff
@@ -374,22 +385,106 @@ static int ipl_alignment_check(void)
 	return -1;
 }
 
+static int initialize_and_check_clock_chip(uint8_t& clock_count)
+{
+	struct pdbg_target *clock_target;
+	enum pdbg_target_status status;
+	uint8_t data;
+	int rc = 0;
+
+	// initialize output variable.
+	clock_count = 0;
+	
+	ipl_log(IPL_DEBUG, "Istep: soft reset clock, and verify clock status register\n");
+	pdbg_for_each_class_target("oscrefclk", clock_target) {
+		status = pdbg_target_probe(clock_target);
+		assert(status == PDBG_TARGET_ENABLED);
+
+		// Resetting the clock chip, so that it will recalibrate input oscilator
+		// signal and identifies bad oscilators.
+		data = OSC_RESET_CMD;
+		rc = i2c_write(clock_target, 0, OSC_CTL_OFFSET, OSC_RESET_CMD_LEN, &data);
+		if(rc){
+			ipl_log(IPL_ERROR, "soft reset command is failed for clock %d, with rc = %d\n",
+				pdbg_target_index(clock_target), rc);
+
+			ipl_error_callback(IPL_ERR_CLK_FAILED);
+
+			return 1;
+		}
+		else{
+			ipl_log(IPL_DEBUG, "soft reset command is successfull for clock %d\n",
+				pdbg_target_index(clock_target));
+		}
+
+		// wait for clock to restart
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+		// Read clock status register to check whether it reports calibration error.
+		// Bit-0 will be set if there is a calibration error.
+		rc = i2c_read(clock_target, 0, OSC_STAT_OFFSET, OSC_STAT_REG_LEN, &data);
+		if(rc){
+			ipl_log(IPL_ERROR, "status register read is failed for clock %d, with rc = %d\n",
+				pdbg_target_index(clock_target), rc);
+
+			ipl_error_callback(IPL_ERR_CLK_FAILED);
+			return 1;
+		}
+		else{
+			ipl_log(IPL_DEBUG, "status register value for clock %d is 0x%2X\n",
+				pdbg_target_index(clock_target), data);
+			if(data & OSC_STAT_ERR_MASK){
+				ipl_log(IPL_ERROR, "Calibration is failed for clock %d, status=0x%2X",
+					pdbg_target_index(clock_target), data);
+				ipl_error_callback(IPL_ERR_CLK_FAILED);
+				return 1;
+			}
+		}
+		clock_count++;
+	}
+    return 0;
+}
+
 static int ipl_set_ref_clock(void)
 {
 	struct pdbg_target *proc;
 	int rc = 0;
+	uint8_t clock_count = 0;
 
 	if (ipl_type() == IPL_TYPE_MPIPL)
 		return -1;
 
 	ipl_log(IPL_INFO, "Istep: set_ref_clock: started\n");
 
+	if(!initialize_and_check_clock_chip(clock_count)){
+		ipl_log(IPL_ERROR, "Clock initialization failed\n");
+		return 1;
+	}
+
+	if(clock_count != 1 && clock_count != 2){
+		ipl_log(IPL_ERROR, "Invalid number (%d) of clock target found\n",
+			clock_count);
+		// This is a software bug. So will not call error callback
+		return 1;
+	}
+
 	pdbg_for_each_class_target("proc", proc) {
 		fapi2::ReturnCode fapirc;
 
 		if (!ipl_is_functional(proc))
 			continue;
 
+		if(clock_count == 2){
```

Comment:
Please include additional comments why this is special action.
Prefer not use hard codeded numbers isnerad use meaning ful contstants.